### PR TITLE
Added code editor to injection interface: the sequel (this time, with CodeMirror!)

### LINKED ISSUE #4995: Code Injection: Line numbers and syntax highlighting
We currently have the lovely [code injection](#1993) feature, which is accessible via the labs page.

In order to get this ready to be fully shipped and no longer classed as a 'labs' feature, we want to upgrade the textarea to have nice syntax highlighting and line numbers for the HTML/JS/CSS code that will likely be pasted in.

There are three main things to do to complete this:

1. decide on what editor library to use to add the new functionality, this could be CodeMirror or suggestions are welcome (CodeMirror is being dropped from the main Ghost editor, so there's no longer a requirement that it be that) 
2. make sure the syntax highlighting and line numbers are styled to look like they fit in the Ghost admin, this might take a bit of toing and froing with the core team 
3. wrap the fancy textarea functionality up into a little ember component so we can reuse it elsewhere in the admin

This might be a bit of a punt for a `beginner` issue, but I'm confident it suits someone new to the Ghost codebase as it's mostly a wiring job :)

### PR DESCRIPTION
**Pull request by acburdine**
This adds code syntax highlighting to the code-injection interface

closes #4995

### CODE REVIEW

**Comment by acburdine** [comment]
@ErisDS so I looked at the vendor.js filesize

vendor.min.js (without ace or codemirror) - 950 kb
vendor.min.js (with ace) - 1.4 MB
vendor.min.js (with codemirror) - 1.1MB

so it still adds 1.5 mb to the files, but it's better than 4.5mb

File: core/client/app/components/gh-cm-editor.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        lineNumbers: true,
+        indentUnit: 4
+    },
+
+    didInsertElement: function () {
+        var options = this.get('options'),
+            self = this,
+            editor;
+        options.mode = this.get('mode');
+        options.theme = this.get('theme');
```
Comment (novaugust) [inline core/client/app/components/gh-cm-editor.js]: You can get at your options a bit more quickly like so, if you like this style:
```js
// codemirror options
mode: 'foo',
theme: 'bar',
lineNumbers: true,
indentUnit: 4,
//...
didInsertElement () {
  var options = this.getProperties('mode', 'theme', 'lineNumbers', 'indentUnit');
//  ....
}
```
not necessary, I just thought it was odd to half-build options outside the function, then add more options inside of the function :)

File: core/client/app/components/gh-cm-editor.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        editor = CodeMirror.fromTextArea(document.getElementById(this.get('name')), options);
```
Comment (novaugust) [inline core/client/app/components/gh-cm-editor.js]: You can get the element for a component via `this.get('element')`, no need for the name property / document.get stuff

Comment (novaugust) [inline core/client/app/components/gh-cm-editor.js]: Ah, I see you want the text area inside the component. You can still drop the `name` work and just do `this.$('textarea')` to get the textarea inside this component

File: core/client/app/components/gh-cm-editor.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        editor.getDoc().setValue(this.get('value'));
```
Comment (novaugust) [inline core/client/app/components/gh-cm-editor.js]: The starting value of codemirror can be passed in with the rest of the options in the `fromTextArea`

**Comment by acburdine** [comment]
@novaugust I don't need to get the component element, I need to get the textarea in the component template, the tagname for the component itself is still a div because of styling things. Although the 'name' property probably needs to be renamed for readability purposes.

**Comment by novaugust** [comment]
I think having options inside of a hash makes it hard to configure those options via the component api. As implemented, you have to set a similarly shaped options object in your controller context and then do

```
//controller
options = {
mode = 
theme =
lines = 
...
}
//template
{{gh-cm-editor options=myOptions}}

// Wouldn't this be nicer?
//template
{{gh-cm-editor lineNumbers=false indentUnit=2}}
```
You could facilitate that by making all the default options into root level properties, and just using `this.getProperties('foo', 'bar', 'baz')` to get your options hash

**Comment by acburdine** [comment]
So for some reason this branch isn't updating........I'm going to close it and reopen another one.....

**Pull request closed without merging**